src/core/thread.d

```diff
@@ -631,7 +633,8 @@ class Thread
         }
 
         // Start thread as non-suspendable, undone in thread_entryPoint
-        atomicStore(m_isInCriticalRegion, true);
+        synchronized (Thread.criticalRegionLock)
+            m_isInCriticalRegion = true;
```

Comment:
> It actually will work, the newly started will leave the critical section right after calling Thread.setThis.

Consider 3 threads:
```
void fun() { new int[100]; }
void test()
{
    auto t1 = new Thread(&fun);
    auto t2 = new Thread(&fun);
    auto t3 = new Thread(&fun);
    t1.start();
    t2.start();
    t3.start();
}
```
Let's take a possible snapshot:
- t1 executes fun, triggers a GC collection and calls thread_suspendAll
- t2 executes some DllMain code, thus still holding the OS lock
- t3 is just being started, with m_isCriticalSection being set, but waiting for the OS lock

If thread_suspendAll has already suspended t2, t3 will never clear m_isCriticalSection. t2 needs to be resumed to get out of the dead lock.